src/main/java/org/terasology/dynamicCities/buildings/BuildingManager.java

```diff
@@ -188,28 +182,47 @@ public void postBegin() {
 
     public Optional<GenericBuildingComponent> getRandomBuildingOfZoneForCulture(String zone, Rect2i shape, CultureComponent cultureComponent) {
         if (buildings.containsKey(zone)) {
-            int max = buildings.get(zone).size();
-            GenericBuildingComponent building;
+            // TODO: needs to be adapted if buildings have a specific spawn chance
+            List<GenericBuildingComponent> availableBuildings =
+                    buildings.get(zone).stream().filter(b -> cultureComponent.availableBuildings.contains(b.name)).collect(Collectors.toList());
+            int numberOfBuildingsForZone = availableBuildings.size();
+
+            GenericBuildingComponent selectedBuilding = null;
             int iter = 0;
-            do {
-                int index = rng.nextInt(max);
-                building = (GenericBuildingComponent) buildings.get(zone).toArray()[index];
+            int index;
+            do { // improve iteration by removing non-fitting buildings from the list
+                index = rng.nextInt(numberOfBuildingsForZone);
+                GenericBuildingComponent candidateBuilding = availableBuildings.get(index);
+                if (isFitting(shape, candidateBuilding)) {
+                    selectedBuilding = candidateBuilding;
+                } else {
+                    availableBuildings.remove(candidateBuilding);
+                    numberOfBuildingsForZone--;
+                }
                 iter++;
-            } while ((!isFitting(shape, building) || !cultureComponent.availableBuildings.contains(building.name)) && iter < 100);
```

Comment:
`!cultureComponent.availableBuildings.contains(building.name)` - this check can be performed before so that you never even consider buildings not available for the culture.